package/libskylark/src/libskylark.c

```diff
@@ -0,0 +1,386 @@
+ * \return  RC
+ */
+void log_client_config(const client_config_t *config)
+{
+  log_debug("client_config: endpoint_url=%s\n", config->endpoint_url);
+  log_debug("client_config: device_uuid=%s\n", config->device_uuid);
+  log_debug("client_config: fd=%d\n", config->fd);
+  log_debug("client_config: num_retries=%d\n", config->num_retries);
+  log_debug("client_config: retry_delay=%d\n", config->retry_delay);
+  log_debug("client_config: retry_max_time=%d\n", config->retry_max_time);
+}
+
+/** Read the Skylark device UUID.
+ *
+ * \param uuid  UUID4 string to return.
+ * \return  RC return code indicating success or failure
+ */
+RC get_device_uuid(char *uuid)
+{
+  if (file_read_string(FILE_SKYLARK_UUID, uuid, UUID4_SIZE) < 0) {
+    return E_GENERIC_ERROR;
+  }
+  return NO_ERROR;
+}
+
+/** Format a device UUID (UUID4) for an HTTP request.
+ *
+ * \param uuid         Device UUID4
+ * \param uuid_header  Formatted string header to return.
+ * \return  RC return code indicating success or failure
+ */
+RC get_device_header(const char *uuid, char *uuid_header)
+{
+  // TODO: Handle error code here
+  sprintf(uuid_header, DEVICE_UID_HEADER_FMT, uuid);
+  return NO_ERROR;
+}
+
+/** Initialize a Skylark client configuration.
+ *
+ * \param config Client configuration to initialize
+ * \return  RC return code indicating success or failure
+ */
+RC init_config(client_config_t *config)
+{
+  RC rc = NO_ERROR;
+  if ((rc = get_device_uuid(config->device_uuid)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_device_header(config->device_uuid, config->device_header)) <
+      NO_ERROR) {
+    return rc;
+  }
+  config->num_retries = 0;
+  config->retry_delay = 0;
+  config->retry_max_time = 0;
+  config->fd = 0;
+  memset(config->endpoint_url, '\0', BUFSIZE);
+  return rc;
+}
+
+/** Setup libcurl global variables.
+ *
+ * \return  RC return code indicating success or failure
+ */
+RC setup_globals(void)
+{
+  CURLcode res = curl_global_init(CURL_GLOBAL_ALL);
+  if (res != CURLE_OK) {
+    log_error("%d %s\n", res, curl_easy_strerror(res));
+    return -E_INITIALIZATION_ERROR;
+  }
+  return NO_ERROR;
+}
+
+/** Cleanup libcurl global variables.
+ *
+ * \return  RC return code indicating success or failure
+ */
+void teardown_globals(void) {
+  curl_global_cleanup();
+}
+
+/**
+ *  Download (i.e., rover) processes.
+ */
+
+/** Callback for the download process: writes to a file descriptor.
+ *
+ * For more details, see:
+ * https://curl.haxx.se/libcurl/c/CURLOPT_WRITEFUNCTION.html
+ *
+ * \param ptr       Pointer to the data received by libcurl
+ * \param size      Item size
+ * \param nmemb     Number of items
+ * \param userdata  Pointer to destination
+ * \return RC return code indicating success or failure
+ */
+static size_t download_callback(char *ptr, size_t size, size_t nmemb,
+                                void *userdata)
+{
+  int *fd = (int *)userdata;
+  ssize_t m = write(*fd, ptr, size * nmemb);
```

Comment:
This definitely needs some error handling/recovery. I saw a few things like this from curl that could be probably dealt with more gracefully.

```
* Illegal or missing hexadecimal sequence in chunked-encoding
* transfer closed with outstanding read data remaining
* ...
```